File: Python Code/ProjectEuler/project_euler_problems/problem141.py
import math
from utility_functions import combinatorics

_limit = int(1e12)


def solution():
    answer = 0
    
    m_limit = int(1e4)
    squares = set()

    for m in range(2, m_limit):
        if m%2==1:
            step = 1;
        else: 
            step = 2
        for r in range(1,m,step):
            if combinatorics.gcd(m,r) != 1:
                    continue
            if m*m*m*r + r*r > _limit:
                break
            for k in range(1,_limit):
                n = k*r*(k*m*m*m+r)
                if n >= _limit:
                    break
                if combinatorics.is_perfect_square(n):
                    if n not in squares:
                        print(k,m,r,n)
                        squares.add(n)
                        answer+=n
    return answer

File: C# Code/EulerProject/Problems/Problem141.cs
using System;
using System.Collections.Generic;
using System.Linq;




namespace EulerProject
{
    public class Problem141
    {
        public static long limit = (long)1e12;
        public static long Solution()
        {
            long solution = 0;
            long mLimit = (long)1e4;
            
            HashSet<long> squares = new HashSet<long>();
            for (long m = 2; m < mLimit; m++)
            {
                long kLimit = (long)Math.Sqrt(limit/(m*m*m)+1);
                var step = 2 - (m % 2);
                for (long r = 1; r < m; r += step)
                {
                    if (UtilityFunctions.Gcd(m, r) != 1)
                        continue;
                    if (m * m * m * r + r * r >= limit)
                        break;
                    for (long k = 1; k <= kLimit; k++)
                    {
                        long n = k * k * m * m * m * r + k * r * r;
                        if (n >= limit)
                            break;
                        if (UtilityFunctions.IsPerfectSquare(n))
                        {
                            if (!squares.Contains(n))
                            {
                                solution += n;
                                squares.Add(n);
                            }
                        }
                    }
                }
            }
            return solution;
        }
    }
}